src/wifi.rs

```diff
@@ -0,0 +1,34 @@
+use anyhow::Result;
+
+use embedded_svc::wifi::{AuthMethod, ClientConfiguration, Configuration};
+use esp_idf_svc::wifi::{BlockingWifi, EspWifi};
+
+#[toml_cfg::toml_config]
+pub struct Config {
+    #[default("")]
+    wifi_ssid: &'static str,
+    #[default("")]
+    wifi_password: &'static str,
+}
+
+pub fn connect_wifi(wifi: &mut BlockingWifi<EspWifi<'static>>) -> Result<()> {
+    let wifi_configuration: Configuration = Configuration::Client(ClientConfiguration {
+        ssid: CONFIG.wifi_ssid.into(),
+        password: CONFIG.wifi_password.into(),
+        auth_method: AuthMethod::WPA2Personal,
+        ..Default::default()
+    });
+
+    wifi.set_configuration(&wifi_configuration)?;
+
+    wifi.start()?;
+    log::info!("Wifi started");
+
+    wifi.connect()?;
+    log::info!("Wifi connected");
+
+    wifi.wait_netif_up()?;
+    log::info!("Wifi netif up");
```

Comment:
What happens if we don't get a connection, is this handled gracefully?

Reply:
Instead of using try/catch (like Python and JS) or returning a tuple with the error (like Go), Rust uses a type called [Result](https://doc.rust-lang.org/rust-by-example/error/result.html). It's an enum containing `Ok(val)` or `Err(error)`.
There are multiple ways to obtain the value within a `Result`, some of which will not be handled gracefully. You can check out that link for more details.

In this case, each of these functions returns a `Result`, and I'm using the [`?` operator](https://doc.rust-lang.org/rust-by-example/error/result/enter_question_mark.html) to return the error if it evaluates to an error, otherwise it continues. All errors are handled this way, so there should be no `panic`s.

Reply:
TLDR: it will fail gracefully.